File: app/src/main/java/com/dreamland/rxcache/policy/LoadFromMemoryFirstPolicy.java
package com.dreamland.rxcache.policy;

import android.content.Context;

import com.dreamland.rxcache.ICacheLoader;
import com.dreamland.rxcache.RxCacheLoaderHelper;
import com.dreamland.rxcache.rxutils.Tuple3;

import rx.Observable;


/**
 * Created by XMD on 2017/6/16.
 */

public class LoadFromMemoryFirstPolicy<T> implements ILoaderPolicy<T> {
    @Override
    public Observable<T> load(Context context,String url, Class<T> clazz, ICacheLoader<T> cacheLoader){
        Tuple3<Context,String,Class<T>> tuple3 = new Tuple3<>(context,url,clazz);
        return cacheLoader.loadFromMemory(Observable.just(tuple3))
                .switchIfEmpty(cacheLoader.loadFromDisk(Observable.just(tuple3)))
                .switchIfEmpty(cacheLoader.loadFromNetwork(Observable.just(tuple3)))
                .switchIfEmpty(new RxCacheLoaderHelper.NoDataObservable<T>().create());
    }


}

File: app/src/main/java/com/dreamland/rxcache/policy/LoadFromMemoryFirstPolicy.kt
package com.dreamland.rxcache.policy

import android.content.Context
import com.dreamland.rxcache.ICacheLoader
import com.dreamland.rxcache.RxCacheLoaderHelper
import com.dreamland.rxcache.rxutils.Tuple4
import rx.Observable
import java.net.URI

/**
 * Created by XMD on 2017/6/20.
 */
class LoadFromMemoryFirstPolicy : ILoaderPolicy {
    override fun <T : Any> load(tuple: Tuple4<Context, URI?, URI?, Class<T>>, iCacheLoader: ICacheLoader) = with(iCacheLoader) {
        loadFromMemory(Observable.just(tuple))
                .switchIfEmpty(loadFromDisk(Observable.just(tuple)))
                .switchIfEmpty(loadFromNetwork(Observable.just(tuple)))
                .switchIfEmpty(loadDefault(Observable.just(tuple)))
                .switchIfEmpty(RxCacheLoaderHelper.NoDataObservable().create())!!
    }
}